kani-compiler/src/kani_middle/codegen_order.rs

```diff
@@ -0,0 +1,65 @@
+
+//! Utilities for optimizing the order of Kani codegen.
+//!
+//! When compiling with more than a single thread, the order in which we codegen harnesses can have
+//! an non-negligable impact on performance. Specifically, if we handle harness that will generate
+//! a lot of code near the end of compilation, the main compiler thread can get stuck waiting for
+//! worker threads to export that code, slowing down overall compilation.
+//!
+//! This module currently provides a simple [MostReachableItems] heuristic to combat that, but more
+//! complex heuristics might be able to improve on this or avoid other kinds of pitfalls.
+
+use crate::{
+    codegen_cprover_gotoc::HarnessWithReachable, kani_middle::transform::BodyTransformation,
+};
+
+/// Orders harnesses within a [CodegenUnit](crate::kani_middle::codegen_units::CodegenUnit) based on
+/// **the raw number of items found during reachability analysis**, putting those with more first.
+///
+/// The number of reachable items seems to be a good proxy for the amount of code we will generate and
+/// thus how long both codegen and the goto file exporting will take. Putting the harnesses that will take
+/// the longest first ensures that
+pub struct MostReachableItems;
+
+impl CodegenHeuristic for MostReachableItems {
+    fn evaluate_harness(harness: &HarnessWithReachable) -> usize {
+        harness.1.reachable.len()
+    }
+}
+
+pub trait CodegenHeuristic {
+    /// Evaluate and rate a harness based on the given heuristic (where *higher is better*).
+    fn evaluate_harness(harness: &HarnessWithReachable) -> usize;
+}
+
+fn reorder_harnesses<'a, H: CodegenHeuristic>(
+    harnesses: &mut Vec<(HarnessWithReachable<'a>, BodyTransformation)>,
+) {
+    // Sort is ascending by default, so `usize::MAX - ...` ensures higher rated harnesses come first.
+    // We don't care about stability, and for cheap heuristic fns like the one for `MostReachableItems`,
+    // caching isn't likely to make a difference.
+    harnesses.sort_unstable_by_key(|(harness, _)| usize::MAX - H::evaluate_harness(harness));
+}
+
+/// Simple trait extender to allow us to call `.apply_...()` on the right kind of iterators.
+/// Could also just be implemented as a function, but this matches the iterator style now used
+/// for reachability in `codegen_crate`.
+pub trait HeuristicOrderable: Iterator {
+    fn apply_ordering_heuristic<T: CodegenHeuristic>(self) -> impl Iterator<Item = Self::Item>;
+}
+
+impl<'a, I> HeuristicOrderable for I
+where
+    I: Iterator<Item = Vec<(HarnessWithReachable<'a>, BodyTransformation)>>,
+{
+    /// Apply an codegen ordering heuristic to an iterator over codegen units.
+    fn apply_ordering_heuristic<H: CodegenHeuristic>(self) -> impl Iterator<Item = I::Item> {
+        // Reorder harnesses within each codegen unit according to `T`.
+        self.map(|mut harnesses| {
+            reorder_harnesses::<H>(&mut harnesses);
+            harnesses
+        })
```

Comment:
```suggestion
        }).collect::<Vec<_>>().into_iter()
```

I could not tell you why tbh, but testing is showing that forcing all of the ordering to happen up front with a collect rather than lazily with a map makes this change go from slightly worse to significantly better.

Reply:
I need to look into why this is, as collecting seems to sneakily guzzle memory (peak goes from normal ~2 GB to ~6 GB), I think bc it requires one transformer per codegen unit to be kept in memory? If this change is doable without that drawback, it may be worthwhile but, if not, it should totally be dropped

Reply:
The cause of the memory overhead is that, now that we are doing reachability in a single pass first, we're inadvertently **storing the reachable call tree for every harness** while we figure out which order to codegen them in. Definitely not ideal as this takes up a ton of space :(. I'll look into how to fix this and if other heuristics could help